ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+            { SmallProofGenerator::RECURSION_FACTOR },
+        >::new();
+        // lagrange table
+        let lagrange_table = LagrangeTable::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+            { SmallProofGenerator::LAGRANGE_LENGTH },
+        >::from(denominator);
+
+        // recursively generate proofs
+        // until there are less than `RECURSION_FACTOR * (RECURSION_FACTOR-1)` many (u,v) field elements
+        // and then do one more iteration
+        // This ensures that after the last iteration,
+        // there are at most RECURSION_FACTOR-1 many real (u,v) field elements
+        // Therefore at least the last field element is a `F:ZERO` filler that can be replaced
+        // Therefore we have space to include the masks during the final proof
+        //
+        // we remark that uv_value.len() will always be at least 1
+        // further, polynomial uv_values.uv[0].0 and .1 will always have
+        // RECURSION_FACTOR many points
+        // since they are arrays of fixed length, i.e. RECURSION_FACTOR,
+        // (which are filled with `F::ZERO` when there are not enough elements).
+        let r_minus_one = SmallProofGenerator::RECURSION_FACTOR - 1;
+        uv_values = loop {
+            let stop = uv_values.len() <= r_minus_one;
+
+            // generate next proof
+            let (uv_values_new, proof_from_left, prover_left_proof) =
+                SmallProofGenerator::compute_next_proof(
+                    ctx,
+                    &mut record_counter,
+                    &lagrange_table,
+                    uv_values.iter(),
+                );
+            // collect proof
+            proofs_from_left.push(proof_from_left);
+            prover_left_proofs.push(prover_left_proof);
+
+            uv_values = uv_values_new;
+
+            if stop {
+                break uv_values;
+            }
+        };
+
+        // generate masks
+        // verifier on the right has p,
+        // therefore the right share is "implicitly sent" to the right ("communicated" via PRSS)
+        let (p_mask_from_left, p_mask_prover): (Fp61BitPrime, Fp61BitPrime) =
+            ctx.prss().generate_fields(record_counter);
+        record_counter += 1;
+        // and verifier on the left has q
+        // therefore the left share is "implicitly sent" to the left (communication via PRSS)
+        let (q_mask_prover, q_mask_from_right): (Fp61BitPrime, Fp61BitPrime) =
+            ctx.prss().generate_fields(record_counter);
+        record_counter += 1;
+
+        // compute final uv values
+        let (u_values, v_values) = &mut uv_values[0];
+        // check that last elements are really Zero
+        debug_assert_eq!(
+            u_values[SmallProofGenerator::RECURSION_FACTOR - 1],
+            Fp61BitPrime::ZERO
+        );
+        debug_assert_eq!(
+            v_values[SmallProofGenerator::RECURSION_FACTOR - 1],
+            Fp61BitPrime::ZERO
+        );
```

Comment:
I think if we make a UVStorage struct like I suggested above, it can keep track of its *populated* length (as opposed it it's allocated length), and we can perform these checks more elegantly, and in a more abstracted way.